xtask/src/test_runner.rs

```diff
@@ -0,0 +1,936 @@
+                out_clone2.lock().unwrap().push('\n');
+            }
+        });
+
+        Ok((child, output, stdout_handle, stderr_handle))
+    }
+
+    /// Runs a command with a timeout, returning the exit code
+    pub fn run_command_with_timeout(&self, cmd: &mut Command, timeout: Duration) -> Result<i32> {
+        log::debug!("Running command: {cmd:?}");
+        self.setup_command(cmd);
+
+        let mut child = cmd.spawn()?;
+        let completed = Arc::new(AtomicBool::new(false));
+        let child_id = child.id();
+        let completed_clone = Arc::clone(&completed);
+
+        let timer = thread::spawn(move || {
+            let interval = Duration::from_millis(100);
+            let mut elapsed = Duration::ZERO;
+
+            while elapsed < timeout {
+                thread::sleep(interval);
+                elapsed += interval;
+                if completed_clone.load(Ordering::SeqCst) {
+                    return;
+                }
+            }
+
+            log::warn!("Command timed out after {timeout:?}, killing process {child_id}");
+            Self::terminate_process(&mut None);
+        });
+
+        let status = match child.wait() {
+            Ok(s) => {
+                completed.store(true, Ordering::SeqCst);
+                s
+            }
+            Err(e) => {
+                completed.store(true, Ordering::SeqCst);
+                thread::sleep(Duration::from_millis(10));
+                return Err(format!("Command execution failed: {e}").into());
+            }
+        };
+
+        let _ = timer.join();
+        let exit_code = status.code().unwrap_or(1);
+        log::debug!("Command exit code: {exit_code}");
+        Ok(exit_code)
+    }
+
+    /// Runs a command for a specified duration, returning whether it terminated
+    /// naturally
+    pub fn run_command_for(&self, cmd: &mut Command, duration: Duration) -> Result<bool> {
+        log::debug!("Running command: {cmd:?}");
+        let mut child = cmd.spawn()?;
+        let start_time = Instant::now();
+        let mut naturally_terminated = false;
+
+        if let Ok(Some(_)) = child.try_wait() {
+            naturally_terminated = true;
+        } else {
+            thread::sleep(duration);
+            if let Ok(Some(_)) = child.try_wait() {
+                naturally_terminated = true;
+            }
+        }
+
+        if !naturally_terminated {
+            log::info!(
+                "Command ran for {:?}, terminating process {}",
+                start_time.elapsed(),
+                child.id()
+            );
+            Self::terminate_process(&mut Some(&mut child));
+        }
+
+        Self::restore_terminal();
+        log::debug!("Command completed after {:?}", start_time.elapsed());
+
+        Ok(naturally_terminated)
+    }
+
+    fn create_espflash_command(&self, args: &[&str]) -> Command {
+        let debug_path = PathBuf::from("target/debug/espflash");
```

Comment:
This can be simplified to `cargo run -p espflash --release --`, I think?

Reply:
I think this would execute `cargo run` on the VMs, which is something we dont want as that would be very slow. We could have a `--local` (or whatever) name that uses  `cargo run` instead of trying to invoke the `espflash` bin on the system `PATH`

Reply:
Can we do what we do in esp-hal, and use a more powerful runner to build the espflash binary for the PR, and then download it onto the VM? We can use the macos self hosted, or even a github runner would be faster than these vms :D

Reply:
Thats what we currently do, we [build `xtask` and `espflash` bins on a gh-runner,](https://github.com/esp-rs/espflash/pull/912/files#diff-3f888d600c3bcb11dd1c4600c03a6249d5741149bbcb7d04bf3e7b037f5cc693R52-R65) download them in the VMs and execute them. But if we try to do `cargo run -p espflash --release --...` on the VM, it will try to buil the binaries again and it will be sloooow

Reply:
Gotcha, so for CI we should build it first and do what you mentioned above, but for local tests we should `cargo run -p espflash --release` for local code.